pkg/syncer/namespace/namespace_downstream_process.go

```diff
@@ -87,8 +87,23 @@ func (c *DownstreamController) process(ctx context.Context, key string) error {
 		return nil
 	}
 	if !exists {
-		logger.Info("deleting downstream namespace because the upstream namespace doesn't exist")
-		return c.deleteDownstreamNamespace(ctx, namespaceName)
+		// Check if the namespace is empty before trying to delete it
+		if empty, err := c.isDowntreamNamespaceEmpty(ctx, downstreamNamespace.GetName()); err != nil {
+			logger.Error(err, "failed to check if downstream namespace is empty")
+			return nil
+		} else if !empty {
+			logger.V(2).Info("downstream namespace is not empty, refusing to delete")
+			return nil
+		}
+		if !c.isMarkedForDeletion(key) {
+			logger.Info("Adding namespace to the delayed delete queue")
+			c.markForDeletion(key)
+		}
+		return nil
+	}
+	// The namespace exists upstream, so we can remove it from the delayed delete queue
+	if c.isMarkedForDeletion(key) {
```

Comment:
I don't think you need to do this test. You should be able to call `cancelDeletion()` directly isn't it ?

Reply:
done